jax/interpreters/numpy_eval.py

```diff
@@ -0,0 +1,76 @@
+# Copyright 2018 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+
+from contextlib import contextmanager
+from typing import Any, Dict, Callable
+
+from jax import config, core
+from jax.interpreters.xla import DeviceArray
+
+np_impl: Dict[core.Primitive, Callable[..., Any]] = {}
+is_tracing = False
+
+class NumpyEvalTrace(core.Trace):
+  def pure(self, x): return x
+  lift = sublift = pure
+
+  def process_primitive(self, primitive, tracers, params):
+    impl = np_impl.get(primitive)
+    if impl is None:
+      raise NotImplementedError(f"No original numpy implementation for "
+                                f"'{primitive}' primitive found.")
+    tracers = [t.copy() if isinstance(t, DeviceArray) else t for t in tracers]
```

Comment:
I don't think this line should be necessary; DeviceArrays support the `__array__` method so that numpy can turn them back into ndarrays at will.

Does it work if you remove this line?

Reply:
Removing this broke `slice` and `update_slice` when applied to operands of type `DeviceArray`. I fixed it with a small workaround that can be removed once `gather_p` is implemented.